File: FE/src/app/services/comments.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { catchError, Subject, throwError } from 'rxjs';
import { environment } from 'src/environments/environment';

import { PagingParameters } from '../models/book-parameter.model';
import { CommentRequest } from '../models/comment-request.model';
import { CommentResponse } from '../models/comment-response.model';
import { PagedRequest } from '../models/paged-request.model';

@Injectable({
  providedIn: 'root'
})
export class CommentsService {
  public pagedCommentsChanged = new Subject<PagedRequest<CommentResponse>>();
  private commentsEndpoint = environment.baseUrl + '/Comments';

  constructor(private http: HttpClient) { }

  public addComment(content: string, bookId: string) {
    const commentRequest = new CommentRequest(content, bookId);
    return this.http.post<CommentResponse>(this.commentsEndpoint, commentRequest).pipe(
      catchError(error => {
        let errorMessage = error.error.message;
        return throwError(() => errorMessage);
      })
    );
  }

  public getCommentsByBookId(bookId: string, params: PagingParameters) {
    let httpParams = new HttpParams();
    httpParams = httpParams.append('pageNumber', params.pageNumber);
    httpParams = httpParams.append('pageSize', params.pageSize);

    return this.http.get<PagedRequest<CommentResponse>>(
      `${this.commentsEndpoint}/Book/${bookId}`, {
      params: httpParams
    }).subscribe(pagedComments => {
      this.pagedCommentsChanged.next(pagedComments);
    });
  }

  public getPendingComments(params: PagingParameters) {
    let httpParams = new HttpParams();
    httpParams = httpParams.append('pageNumber', params.pageNumber);
    httpParams = httpParams.append('pageSize', params.pageSize);

    return this.http.get<PagedRequest<CommentResponse>>(
      `${this.commentsEndpoint}/Pending/`, {
      params: httpParams
    }).subscribe(pagedComments => {
      this.pagedCommentsChanged.next(pagedComments);
    });
  }

  public approveComment(commentId: string) {
    return this.http.put<CommentResponse>(
      `${this.commentsEndpoint}/Approve/${commentId}`, {}).pipe(catchError(error => {
        let errorMessage = error.error.message;
        return throwError(() => errorMessage);
      }))
  }

  public rejectComment(commentId: string) {
    return this.http.put<CommentResponse>(
      `${this.commentsEndpoint}/Reject/${commentId}`, {}).pipe(catchError(error => {
        let errorMessage = error.error.message;
        return throwError(() => errorMessage);
      }))
  }
}

File: BE/BusinessLayer/Services/CommentsService.cs
using BusinessLayer.Interfaces.Comments;
using BusinessLayer.Models;

namespace BusinessLayer.Services
{
    public class CommentsService : ICommentsService
    {
        private readonly ICommentsRepository commentsRepository;

        public CommentsService(ICommentsRepository commentsRepository)
        {
            this.commentsRepository = commentsRepository;
        }

        public async Task<Comment> AddComment(Comment comment)
            => await commentsRepository.AddComment(comment);

        public async Task<Comment> ApproveComment(Guid commentId)
            => await commentsRepository.ApproveComment(commentId);

        public async Task<Comment> RejectComment(Guid commentId)
            => await commentsRepository.RejectComment(commentId);

        public PagedList<Comment> GetCommentsByBookId(Guid bookId, PagingParameters pagingParameters)
            => commentsRepository.GetCommentsByBookId(bookId, pagingParameters);

        public PagedList<Comment> GetCommentsByUserId(Guid userId, PagingParameters pagingParameters)
            => commentsRepository.GetCommentsByUserId(userId, pagingParameters);

        public PagedList<Comment> GetPendingComments(PagingParameters pagingParameters)
            => commentsRepository.GetPendingComments(pagingParameters);
    }
}